### DISCUSSION TITLE
Use Case: Text Translation using both Autographa and Paratext

### DISCUSSION THREAD

**Discussion by joelthe1** [bible-technology/scripture-burrito#240] [General]
This is a real use-case of how some teams use Autographa/translationCore/Paratext to complete a translation project which serves as a good lens for understanding the viability of SB. The translation proceeds in phases and may use one or more different application to complete each phase making it important that the interchange is lossless and valid.
1. The MTT (Mother Tongue Translator) uses Autographa to start a text translation focused mainly on drafting and preliminary syntactic/spelling checks. The project is synced on Door43. 
2. The MTT then hands the above completed draft over to quality checkers who may use Paratext and/or translationCore to complete multiple phases of checking all while syncing the SB with Paratext and/or Door43 Registry.
3. The final version may then be used by publishing software/apps which may be in the original text translation SB or converted to a more format closer to what the consumer understands.

**Comment by jtauber** [comment]
Is it important that the burritos are treated as different versions of the same thing throughout this? Or are they successive new burritos declared to be derived from a previous one?

**Reply by joelthe1** [reply]
The way I see it, given the fact that the changes across different stages maybe substantial and that they represent a different quality standard (meeting some predefined specification), seeing the each stage deriving itself from the previous fits better.

This does raise a concern on how to disambiguate these two concepts better. In some ways, we have to trust the users but ideally there would be only one way to do it (at the cost of some loss of information).

**Comment by jag3773** [comment]
In this scenario, I think the import thing is that each tool that is importing a given burrito knows what project to associate it with internally.  I'll put some more details in here, filling out Joel's example above. Note that I'll use numbers to indicate the different burritos that would be exchanged for this single project.

1. User starts a drafting in Autographa, syncing natively with git.door43.org
2. User exports project as a SB#1 and imports it to Paratext (local process, on the same computer)
3. Paratext must either a) create a new project for this, or b) associate it with an existing project (at this point, option a since this is a new project).
4. User now exports from Paratext SB#2 and imports to translationCore.
5. translationCore must either a) create a new project for this, or b) associate with an existing project. (at this point, either option should be possible, since translationCore and Autographa can both sync with git.door43.org, it should be possible for translationCore to automatically identify the project on git.door43.org)
6. Use now exports their updated translation from translationCore SB#3 and imports it back into Paratext
7. Paratext must either a) create a new project for this, or b) associate it with an existing project (at this point, option b is what the user wants, they want to update the existing project).

If our `id` section is allowed to keep track of the project's identifiers as the files are passed around it should be easy for any tool to associate the burritos on import with the correct project in their tool or ecosystem. If we do not support this function then it seems like we'll have to prompt the user for more information on import (undesired).

Note that the use case for derived (previously [variant](https://docs.burrito.bible/en/v0.3.0-alpha.2/glossary.html?highlight=variants)) burritos is a programmatically derived subset of a project, not merely a successive instantiation of a project.

**Comment by jonathanrobie** [comment]
> In this scenario, I think the important thing is that each tool that is importing a given burrito knows what project to associate it with internally.

That's good if we can do it automatically. I'm not sure it is possible in general.  How does a tool know whether the user wants to import into an existing project or create a new project?  If the user wants to import into an existing project, how does the tool know whether to merge or overwrite changes?  I suspect a tool will have to ask the user some questions.

Still, identifiers can be useful even if we do need to ask the user questions.   "Do you want to update this project or create a new one?"

> If our `id` section is allowed to keep track of the project's identifiers as the files are passed around it should be easy for any tool to associate the burritos on import with the correct project in their tool or ecosystem. If we do not support this function then it seems like we'll have to prompt the user for more information on import (undesired).

This basically allows each tool to ask the questions I mentioned above.  It doesn't mean the tool automatically knows what the user wants.  But it could be useful.

To be useful when a given resource is managed in two different ecosystems, we probably have to say that servers need to preserve all known resource identifiers and provide them when a resource is exported.  If that's not required, we won't have interoperability at that level.   Even if we do require it, we will not get consistent versioning across the ecosystems, that is up to the user.

To me, the right solution is to solve the disconnect between the ecosystems so that there is a single identifier and consistent versioning.  That's not something the file transfer format will do for us.  

If we don't do that, I think it's basically a cost/benefit question.  How many clients will actually be doing this across multiple ecosystems?  To me, this feels like adding complexity for a solution that doesn't actually get us that far in solving the real problem.

Or am I missing something?

**Reply by joelthe1** [reply]
I agree that it doesn't solve the bigger problem but like you pointed out from a cost/benefit point of view, I think there is little to lose in requiring clients to not touch others IDs. While on the contrary it does give _some_ benefit (possible connection to an existant project).

Do you see any reasons for keeping the IDs around in the burrito that may be detrimental?

**Reply by jag3773** [reply]
I agree @jonathanrobie and I also agree @joelthe1 -- doesn't seem harmful to persist IDs and will likely be helpful.

**Comment by jonathanrobie** [comment]
Both @jessegriffin and @joelthe1 seem to be thinking of the burrito as a
persistent instance on disk.  I am thinking of a Scripture Burrito as an
exchange format. I don't know what it means to "delete" things out of an
exchange format - is there an expectation that Paratext needs to store
everything in a Burrito and use it to create new ones?  If so, how does
Paratext know which revisions each piece of metadata still applies to?

I think this is really a question of use and scope.  I am guessing that
perhaps you are using Scripture Burrito on disk as your internal format,
then committing it to GitHub?  Do unfoldingWord and Autographa share the
same GitHub instance?  If so, I gather unfoldingWord would simply
redistribute the Autographa instance, without changing its metadata?  I'm
just guessing.  If both the unfoldingWord and Autographa are updating this
data, how do you define consistency?  I assume that would get you into all
the "editing burrito" scenarios that we have taken out of scope.

I think we have agreed that the Scripture Burrito standard defines an
exchange format.  If any software happens to use the same format
internally, cool, but our standard does not need to define what that means,
what consistency constraints apply, and how multiple programs can handle
that consistently on a GItHub repo.   If this kind of software needs
additional information of any kind, I think that's what extensibility is
for.  But we haven't even begun to design this as a shared persistence
format, and if we do that, there are some database considerations I would
want to look at with use cases and requirements.  For instance, suppose
Autographa, Paratext Lite, and Paratext all share the same projects with
the Paratext servers, whose metadata goes in the persistent instance?  I
think that raises some complexities, and I don't think it is part of the
Minimum Required to Declare Victory.  For that, we need only an exchange
format.

I'm not sure, though.  I do think spelling out the use case in more detail
would be helpful.  For instance, if Autographa gets a Scripture Burrito
from Paratext, it is probably a Burrito that Paratext created, not the same
one Autographa sent. Can you sketch out exactly what is exchanged with
GitHub, unfoldingWord, and Paratext in this scenario, perhaps especially
focusing on what you want Paratext to give you in a Scripture Burrito
compared to what Autographa sent?

**Reply by jag3773** [reply]
I think the issue is better cast in light of our definition of "lossless" metadata. I don't think it relates at all to on disk format but rather to what metadata needs to persist.

For example, since lossless metadata is a goal, I could make the argument that not keeping track of IDs is a violation. At the same time, we've agreed that history is not in scope for a Burrito, so you can make the argument (as you are @jonathanrobie ), that any notion of "persisting" anything is a violation.

It could be that we're trying to do too much with the IDs, but it could also be that this would be a helpful feature.

As @jonathanrobie described above, our users will have to do some project identification on import of a burrito, I'm trying to make sure that process can be as simple as possible.

**Reply by jonathanrobie** [reply]
I think we probably need to be more concrete about what we mean by "lossless" metadata, especially since we have defined the scope and architecture more clearly than we had when we set that goal.  Lossless compared to what?  

Here are the two statements I see in our spec:

> Our goal is lossless portability of Scripture-related metadata and data between translation and publication users, applications and ecosystems.

What does that mean?  If a Paratext project is uploaded to DBL and published, that publication has lost a lot of information contained in the Paratext project.   How would the meaning of that requirement change if we deleted the word "lossless"?

> Scripture Burrito is intended to allow lossless roundtripping of projects between ecosystems. This depends to some extent on references to ecosystem servers that enable reconnection with different ecosystem-specific contexts.

That's a lot more than a file format is going to do.  It's probably a lot easier to get our tools to work together in one ecosystem than to create a metasystem that mediates among them as though they were in one ecosystem.  If we have the level of cooperation needed to create that metasystem, we have the level of cooperation we need to make our tools work together in a simpler way than that.  I think we should pursue that, but it's beyond the scope of what Scripture Burrito can do for us.

And what does "lossless" mean here?  Again, there's a lot of metadata associated with a translation project that is not contained in a Burrito, including version history, user information, etc.

I think we probably mean something more specific than "lossless roundtripping".  We should probably specify what we do mean as concretely as possible.

**Reply by jag3773** [reply]
The "lossless roundtripping" refers to the information in the Burrito, not to other information that's outside the SB spec.

**Reply by jonathanrobie** [reply]
> The "lossless roundtripping" refers to the information in the Burrito, not to other information that's outside the SB spec.

So how do I interpret "lossless roundtripping of projects between ecosystems"?  Presumably, if system `a` sends burrito `sba` to systems `b` and `c`, they do not send precisely the same burrito `sba` back to system `a`.  

How is that sentence different if we delete "between ecosystems"?  Are we merely saying that a system can write to a burrito and read it back, whether or not there are other systems in between?  If so, what does 'lossless' mean here?  I assume most systems have additional information about a project that is not written to the burrito.

What exactly does this mean, and how would I write a test to make sure that we have achieved the goal?

**Comment by jag3773** [comment]
Rewrite

>Scripture Burrito is intended to allow lossless roundtripping of projects between ecosystems.

To focus on use cases and build tests for this into first Autograph/Paratext use case. Provide advisory information what should be included to successfully transfer a project from one tool to another via a SB.
